BringYourOwnBoard.md

```diff
@@ -0,0 +1,126 @@
+
+# Recipes
+
+## recipes-bsp 
+
+### What does it do? 
+
+Within our project `recipes-bsp` is used for appending instructions to the bootloader that allow SwUpdate to set which partition to boot into after completing an image update (eg if we've completed an update while running on `/dev/mmcblk0p0` then when the device restarts we boot into `/dev/mmcblk0p1` which contains the new version of the image). For our purposes we can use [rpi-u-boot-scr.bbappend](./recipes-bsp/rpi-u-boot-scr/rpi-u-boot-scr.bbappend) to add the [boot.cmd.in](./recipes-bsp/rpi-u-boot-scr/files/boot.cmd.in) file to the Raspberry Pi 4's boot process which instructs the device to do the A/B switching. 
+
+Within this layer we also edit u-boot to use our own firmware environment configuration and append a special patch for a gcc issue that RaspberryPi has been dealing with. 
+### What do I need to think about? 
+
+You must talk to your BSP provider and determine what bootloader they provide as apart of the BSP. Likely you will need to create a more durable u-boot script that handles detecting if an update occurred, determines what partition to load into, and provide the logic to do that switching. 
+
+For instance you can look at this snippet: 
+```u-boot
+fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs
+fatload mmc 0:1 ${kernel_addr_r} @@KERNEL_IMAGETYPE@@
+if env exists rpipart;then setenv bootargs ${bootargs} root=/dev/mmcblk0p${rpipart}; fi
+if test ! -e mmc 0:1 uboot.env; then saveenv; fi;
+@@KERNEL_BOOTCMD@@ ${kernel_addr_r} - ${fdt_addr}
+```
+
+Within the [recipes-core] layer we create three partitions and mount them in the `fstab` : `/dev/mmcblk0p2` (the 'A' partition), `/dev/mmcblk0p3` (the 'B' partition), and `/dev/mmcblk0p4` (the "adu" partition). When an update occurs we write the partition number to boot into next to the u-boot environment section setup using `fw-env.config` which is then used by this u-boot script to determine what partition to load into on the next boot. 
+
+For a production device the core idea of using a flag to say what to load into is completely acceptable, however there are some safety checks you should add if you want your device to be secure and durable. 
+
+First you should also write the hash of the image to be booted into. Then u-boot should hash the partition you're attempting to write into (at least include a CRC checksum) and determine if the partition is what you expect. This prevents direct flash injection attacks and keeps your device from running unsafe code. 
+
+Second you want to add a method for handling a kernel panic. That is if you've completed an update and somehow the partition you're booting into isn't bootable you have a way to fall back from the B side to the A side or vice versa. You can accomplish that using u-boot flags and some logic around what to do when. That is essential for preventing your nice, expensive devices from turning into bricks. 
```

Comment:
Re this: _That is essential for preventing your nice, expensive devices from turning into bricks._ I love it personally, it made me smile. I am not sure if that tone is typical for this type of documentation from Microsoft, though? I think a challenge for ADU as we work with the open-source community is to appear more "human" (not a thing MS has historically been super good at), so if this advances that goal, I'm fine with it. I'm just comparing it to the editorial reviews I've had from ADU content on learn.ms.com, and they are a little more protective of tone.

Reply:
I'm fine with whatever messaging you think is best. I wrote this like I would write any other piece of documentation I was handing off to a customer. I've found that adding a little bit of personality tends to make the documentation more palatable, but I'll be the first one to admit I tend to lean heavily towards a casual, friendly tone, like I would use with coworkers, rather than a regimented one that many view as the "correct" corporate response. 

Just depends on the culture you want to cultivate GitHub. Either work.